Mirror climb

ItsDang opened the pull request:
For #48, Added mirror climb button to drop down menu. It acts like a toggle. Currently it resets back to not mirrored if you refresh or switch searches. Also logs mirrored send, I've tested send.

gardaholm commented:
Nice, had a quick play today – logging seems to work fine 👌
Only thing missing is that you can't send the mirrored version to the board via bluetooth. Tried it on the TB2 but I only get the original layout. Would be great to have this addition to.

lemeryfertitta commented on climbdex/templates/results.html.j2:
```diff
@@ -94,6 +94,11 @@
             </button>
             <ul class="dropdown-menu dropdown-menu-end">
+              <li>
+                <button id="button-mirror" type="button" class="dropdown-item btn" data-bs-toggle="button" onclick="mirrorClimb()">
```
I think we should hide this when the board is not symmetrical (e.g. Kilter, Tension Spray). One idea would just be to query the DOM for `data-mirror-id` and hide the button when no results are found. There may be a better way as well.

I'm okay with the list item for now, but I could see a button being more desirable eventually. Probably better to save for a more complete UI redesign as more buttons/list items are added.

lemeryfertitta reviewed (changes requested):
Thanks @ItsDang! Tested and looks great to me, nice clean implementation as well. Just one change needed to hide (or disable) the button on non-symmetrical layouts.

ItsDang commented:
Pulling whether layout is mirrored from db and using to show or hide Mirror Climb button. Updated Bluetooth to pass mirrored, will test on a TB2 at another gym when I can. Refactored whether the the Mirror Climb button is pressed or not.

ItsDang commented:
Realized I mixed up the mirrored sending to board over Bluetooth, was able to test on tb2 at another gym.

ItsDang commented:
Commited the change to light up  the right version, should be ready for review.

lemeryfertitta commented on climbdex/db.py:
```diff
@@ -105,6 +105,8 @@
             climbs.description,
             climbs.frames,
+            layouts.is_mirrored,
+            climbs.frames,
```
Seems like this line is a dupe of L107, but probably don't need to modify the search query at all for this changeset as described in other comment.

lemeryfertitta commented on climbdex/static/js/results.js:
```diff
@@ -3,10 +3,32 @@ const colorMap = colors.reduce((acc, colorRow) => {
   name,
   frames,
+  is_mirrored,
+  mirroredFrames,
```
Seems like mirroredFrames is unused and is just being overwritten later in the method.

lemeryfertitta commented on climbdex/static/js/results.js:
```diff
@@ -289,20 +332,22 @@ function drawResultsPage(results, pageNumber, pageSize, resultsCount) {
       description,
       frames,
+      is_mirrored,
+      mirroredFrames,
```
It seems like this is just `frames` from the SQL query results, not `mirroredFrames` - it also appears to be extra, given that frames are available above.

lemeryfertitta commented on climbdex/static/js/results.js:
```diff
@@ -87,14 +117,27 @@ function drawClimb(
       ledColors
     );
-    illuminateClimb(board, bluetoothPacket);
+    const mirroredBluetoothPacket = getBluetoothPacket(
```
Rather than calculate two separate packets and only use one, we can just do the following:

```javascript
const bluetoothPacket = getBluetoothPacket(
      isMirroredMode() ? mirroredFrames : frames,
      placementPositions,
      ledColors
    );
```

ItsDang commented:
Thanks for comments, it's cleaner. I've implemented and tested the mirrored button except for the Bluetooth on an actual board.

lemeryfertitta reviewed (approved):
Awesome, thanks for this very useful feature. I won't be near a board to test the bluetooth anytime soon, but the logic looks correct and we can always push another update if needed. Will get this merged and deployed now

gardaholm commented:
Nice and really needed addition! 
Will do some tests tomorrow

ItsDang commented:
@lemeryfertitta, Thanks for the help cleaning up the code! And thanks for testing @gardaholm.

gardaholm commented:
Didn't had any troubles with mirroring and lighting up the routes. Also behaves as I would expect it (similar to the app). If you mirror once, all upcoming routes show up as mirrored. Really nice!

Only issue I had is documented here https://github.com/lemeryfertitta/Climbdex/issues/45#issuecomment-2337255576, but thats part of the overall bluetoot discussion I guess
